File: 2017/test/day_09_test.exs
defmodule AdventOfCode2017.Day09Test do
  use ExUnit.Case

  import AdventOfCode2017.Day09
  import AdventOfCode2017.TestHelper

  test "part1" do
    assert part1(sample_input_1()) == 1
    assert part1(sample_input_2()) == 6
    assert part1(sample_input_3()) == 5
    assert part1(sample_input_4()) == 16
    assert part1(sample_input_5()) == 1
    assert part1(sample_input_6()) == 9
    assert part1(sample_input_7()) == 9
    assert part1(sample_input_8()) == 3
    assert part1(input("test/input/day_09.txt")) == 20530
  end

  test "part2" do
    assert part2(sample_input_1_2()) == 0
    assert part2(sample_input_2_2()) == 17
    assert part2(sample_input_3_2()) == 3
    assert part2(sample_input_4_2()) == 2
    assert part2(sample_input_5_2()) == 0
    assert part2(sample_input_6_2()) == 0
    assert part2(sample_input_7_2()) == 10
    assert part2(input("test/input/day_09.txt")) == 9978
  end

  defp sample_input_1, do: "{}"
  defp sample_input_2, do: "{{{}}}"
  defp sample_input_3, do: "{{},{}}"
  defp sample_input_4, do: "{{{},{},{{}}}}"
  defp sample_input_5, do: "{<a>,<a>,<a>,<a>}"
  defp sample_input_6, do: "{{<ab>},{<ab>},{<ab>},{<ab>}}"
  defp sample_input_7, do: "{{<!!>},{<!!>},{<!!>},{<!!>}}"
  defp sample_input_8, do: "{{<a!>},{<a!>},{<a!>},{<ab>}}"

  defp sample_input_1_2, do: "<>"
  defp sample_input_2_2, do: "<random characters>"
  defp sample_input_3_2, do: "<<<<>"
  defp sample_input_4_2, do: "<{!>}>"
  defp sample_input_5_2, do: "<!!>"
  defp sample_input_6_2, do: "<!!!>>"
  defp sample_input_7_2, do: "<{o\"i!a,<{i<a>"
end

File: 2018/test/day_09_test.rb
require "test_helper"
require "day_09"

class Day09Test < Minitest::Test
  def setup
    @day_09 = Day09.new
  end

  def test_part1
    assert_equal 32, @day_09.part1(sample_input_1)
    assert_equal 8317, @day_09.part1(sample_input_2)
    assert_equal 146373, @day_09.part1(sample_input_3)
    assert_equal 2764, @day_09.part1(sample_input_4)
    assert_equal 54718, @day_09.part1(sample_input_5)
    assert_equal 37305, @day_09.part1(sample_input_6)
    assert_equal 386151, @day_09.part1(input("day_09.txt"))
  end

  def test_part2
    assert_equal 3211264152, @day_09.part2(input("day_09.txt"))
  end

  def test_parse
    assert_equal [9, 25], @day_09.parse(sample_input_1)
    assert_equal [459, 71790], @day_09.parse(input("day_09.txt"))
  end

  def sample_input_1
    "9 players; last marble is worth 25 points"
  end

  def sample_input_2
    "10 players; last marble is worth 1618 points"
  end

  def sample_input_3
    "13 players; last marble is worth 7999 points"
  end

  def sample_input_4
    "17 players; last marble is worth 1104 points"
  end

  def sample_input_5
    "21 players; last marble is worth 6111 points"
  end

  def sample_input_6
    "30 players; last marble is worth 5807 points"
  end
end
